# Selected papers - for literature review sessions 

| Day&emsp;| Topic | Papers | 
|------|-----------------------------|--|
| Day 1 | Assembly           | [Minimap and miniasm: fast mapping and de novo assembly for noisy long sequences][1] <br> [The present and future of de novo whole-genome assembly][2] |
| Day 2 | Indexed Alignment  | [Genome structural variation discovery and genotyping][3] <br> [Fast and accurate short read alignment with Burrows-Wheeler transform][4] |
| Day 4 | Pan-genomics       | [Computational pan-genomics: status, promises and challenges][5] <br> [Variation graph toolkit improves read mapping by representing genetic variation in the reference][6] |

# Some other recent publications
A non-exhaustive list of suggested papers 



### Assembly 
- [(2016) Minkin, I., et al. "TwoPaCo: An efficient algorithm to build the compacted de Bruijn graph from many complete genomes." Bioinformatics 33.24 pp. 4024-4032][7]
- [(2017) Koren, Sergey, et al. "Canu: scalable and accurate long-read assembly via adaptive k-mer weighting and repeat separation." Genome research 27.5 pp. 722-736][8] 
- [(2017) Wick, Ryan R., et al. "Unicycler: resolving bacterial genome assemblies from short and long sequencing reads." PLoS computational biology 13.6][9] 
- [(2017) Baaijens, Jasmijn A., et al. "De novo assembly of viral quasispecies using overlap graphs." Genome research 27.5 pp. 835-848][10] 
- [(2019) Kolmogorov, M., et al. "Assembly of long, error-prone reads using repeat graphs." Nature biotechnology 37.5 pp. 540.][11]

### Alignment
- [(2016) Liu, B., et al. "deBGA: read alignment with de Bruijn graph-based seed and extension." Bioinformatics 32.21 pp. 3224-3232][12]
- [(2017) Liu, B., et al. "LAMSA: fast split read alignment with long approximate matches." Bioinformatics 33.2 pp. 192-201][13]

### Pan-genomics 
- [(2016) Sheikhizadeh, S., et al. "PanTools: representation, storage and exploration of pan-genomic data." Bioinformatics 32.17 pp. 487-493.][14] 
- [(2017) Eggertsson, H. P., et al. "Graphtyper enables population-scale genotyping using pangenome graphs." Nature genetics 49.11 pp. 1654] [15]
- [(2018) Garrison, E., et al. "Variation graph toolkit improves read mapping by representing genetic variation in the reference." Nature biotechnology 36.9 pp. 875-879][16]
- [(2018) Ding, W., et al. "panX: pan-genome analysis and exploration." Nucleic acids research 46.1 pp.][17]
- [(2018) Jandrasits, C., et al. "seq-seq-pan: Building a computational pan-genome data structure on whole genome alignment." BMC genomics 19.1 pp. 47][18]



[1]: https://academic.oup.com/bioinformatics/article/32/14/2103/1742895
[2]: https://www.ncbi.nlm.nih.gov/pubmed/27742661
[3]: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4108431/pdf/nihms-605372.pdf
[4]: https://www.ncbi.nlm.nih.gov/pubmed/19451168
[5]: https://www.ncbi.nlm.nih.gov/pubmed/27769991
[6]: https://www.ncbi.nlm.nih.gov/pubmed/30125266

[7]: https://doi.org/10.1093/bioinformatics/btw609
[8]: https://doi.org/10.1101/gr.215087.116
[8]: https://doi.org/10.1101/gr.215038.116
[9]: https://doi.org/10.1371/journal.pcbi.1005595
[10]: https://doi.org/10.1101/gr.215038.116
[11]: https://doi.org/10.1038/s41587-019-0072-8
[12]: https://doi.org/10.1093/bioinformatics/btw371
[13]: https://doi.org/10.1093/bioinformatics/btw594
[14]: https://doi.org/10.1093/bioinformatics/btw455
[15]: https://doi.org/10.1038/ng.3964
[16]: https://doi.org/10.1038/nbt.4227
[17]: https://doi.org/10.1093/nar/gkx977
[18]: https://doi.org/10.1186/s12864-017-4401-3

